src/box/index.cc

```diff
@@ -690,49 +681,23 @@ iterator_position_unpack(const char *packed_pos, const char *packed_pos_end,
 	*pos = NULL;
 	*pos_end = NULL;
 	*part_count = 0;
-	uint32_t map_len = 0;
+	int decoded_size = base64_decode(packed_pos,
+					 packed_pos_end - packed_pos,
+					 buf, buf_end - buf);
+	const char *decoded_pos = buf;
+	const char *decoded_pos_end = buf + decoded_size;
 	const char *array = NULL;
-	const char *packed_pos_check = packed_pos;
-	bool has_position = false;
-	if (mp_check(&packed_pos_check, packed_pos_end) != 0)
+	const char *decoded_pos_check = decoded_pos;
+	if (mp_check(&decoded_pos_check, decoded_pos_end) != 0)
 		goto fail;
-	if (packed_pos_check != packed_pos_end)
-		goto fail;
-	if (mp_typeof(*packed_pos) != MP_BIN)
-		goto fail;
-	mp_decode_binl(&packed_pos);
-	packed_pos_check = packed_pos;
-	if (mp_check(&packed_pos_check, packed_pos_end) != 0)
-		goto fail;
-	if (mp_typeof(*packed_pos) != MP_MAP)
-		goto fail;
-	map_len = mp_decode_map(&packed_pos);
-	/* Cannot be empty. */
-	if (map_len < 1)
-		goto fail;
-	for (uint32_t i = 0; i < map_len; ++i) {
-		if (mp_typeof(*packed_pos) != MP_UINT)
-			goto fail;
-		uint32_t key = mp_decode_uint(&packed_pos);
-		switch (key) {
-		case ITERATOR_POSITION_KEY:
-			if (mp_typeof(*packed_pos) != MP_ARRAY)
-				goto fail;
-			has_position = true;
-			array = packed_pos;
-			*part_count = mp_decode_array(&packed_pos);
-			*pos = packed_pos;
-			mp_next(&array);
-			packed_pos = array;
-			*pos_end = packed_pos;
-			break;
-		default:
-			/* Ignore unknown keys for forward compatibility. */
-			break;
-		}
-	}
-	if (!has_position)
+	if (mp_typeof(*decoded_pos) != MP_ARRAY)
 		goto fail;
+	array = decoded_pos;
+	*part_count = mp_decode_array(&decoded_pos);
+	*pos = decoded_pos;
+	mp_next(&array);
```

Comment:
There's no need to use `mp_next` here - `mp_check` called above already advances the position to the data end. We should probably also check that the advanced position coincides with the end of decoded data (i.e. there's no garbage after the MsgPack array).

Reply:
Done